Adding the v2/Teams API

vandr0iy opened the pull request:
Hi! I've noted that in this repo there's no API for v2/Teams, so I added all the GET calls - and also some of the other ones. Please, be kind: this is my literal first OSS contribution in typescript 😅

vandr0iy committed:
feat: first draft of the GetAllTeams v2 API implementation

vandr0iy committed:
feat: adding 2 more GET calls for the Teams API

vandr0iy committed:
doc: updating the docs + adding invocation from mod.ts

vandr0iy committed:
feat: adding more get/list API calls

vandr0iy committed:
feat: adding deleteTeam and addUserToTeam API calls

vandr0iy committed:
feat: add createTeam API call

vandr0iy committed:
fix: renaming the createTeam API call

vandr0iy committed:
fix: wrong endpoint for createTeam. I should stop copypasting

vandr0iy committed:
fix: linter

vandr0iy committed:
fix: chaging the returned values type

danopia commented:
Hello, thanks for the work. I am on vacation currently but can check in for a review in the next week. I assume you've already used the new APIs and they work for your own needs?

vandr0iy commented:
Hi @danopia! basically, I was writing a script that's using the Teams and the Users API - however, when I found this package I only found the latter being implemented - so I added the former. 

While I wrote the script I was able to successfully invoke all of the Teams API calls at least once (+ that one v2/users call that I added myself). Does that answer your question?

danopia commented on v2/teams.ts:
```diff
@@ -0,0 +1,157 @@
+// Common API client contract
+interface ApiClient {
+  fetchJson(opts: {
+    method: "GET" | "POST" | "DELETE";
```
It looks like this is the first DELETE call in this project, and so the new method needs to be also added to `client.ts`:
https://github.com/cloudydeno/deno-datadog_api/blob/657811ec549cd4c16a5cdb096c07118aff45244e/client.ts#L40

danopia commented on v2/lib/identity.ts:
```diff
@@ -10,10 +10,37 @@ export interface ResultPage<T> {
 }
 
-export interface Included {
+export interface TeamsResultPage<T> {
```
Did datadog really need to add another pagination structure within v2 😅 LGTM

danopia commented on v2/teams.ts:
```diff
@@ -0,0 +1,157 @@
+  }
+
+  /** Create a team */
+  async createTeam(name: string): Promise<string> {
+    const words = [...name.toLowerCase().matchAll(/[a-z0-9]+/g)].map((x) => x[0])
+    const handle = words.join("-")
```
This is slightly magic. Does this string manipulation follow a particular recommendation from datadog?
Continued:

vandr0iy replied:
oof. I'm afraid I leaked out some business logic from my script here 😅 refactoring in progress

danopia replied:
I've resolved this by replacing the parameter with an object so that name and handle are provided explicitly.

danopia commented on v2/teams.ts:
```diff
@@ -0,0 +1,157 @@
+      body: {
+        data: {
+          type: "team",
+          attributes: {
+            name: name,
+            handle: handle
+          },
```
I see that there are additional possible [attributes for this API](https://docs.datadoghq.com/api/latest/teams/#create-a-team), so perhaps we can accept a second optional parameter `extraAttributes` that lets the caller override `handle` and/or pass more attributes for the team.

```suggestion
          attributes: {
            name: name,
            handle: handle,
            ...extraAttributes,
          },
```

danopia reviewed (commented):
I have two specific requests on this

danopia commented:
hey, checking back in because I'm going thru all my modules to switch them over to JSR. Do you have more commits to push here? I could also merge and finish the nits separately..

danopia committed:
Change createTeam signature, add trailing commas

danopia committed:
Follow renaming into v2/roles.ts

danopia committed:
Merge branch 'main' into main
